File: c/leet_code/42.c
/**
 * Function to calculate the total amount of trapped rainwater.
 * This function uses the two-pointer technique to find the trapped rainwater.
 *
 * @param height An array of non-negative integers representing the height of
 * each bar.
 * @param heightSize The size of the height array.
 * @return The total amount of rainwater that can be trapped.
 */
int trap(int *height, int heightSize) {
  // Initialize the variables
  int trap = 0, low, high, curr;
  low = 0;
  high = heightSize - 1;

  // Loop until low pointer is less than high pointer
  while (low < high) {
    // If the height at low pointer is less than or equal to the height at high
    // pointer
    if (height[low] <= height[high]) {
      curr = low;
      // Loop until the height at low pointer is less than the height at curr
      // pointer
      while (height[++low] < height[curr]) {
        // Add the difference of height at curr pointer and height at low
        // pointer to trap
        trap += height[curr] - height[low];
      }
    } else {
      curr = high;
      // Loop until the height at high pointer is less than the height at curr
      // pointer
      while (height[--high] < height[curr]) {
        // Add the difference of height at curr pointer and height at high
        // pointer to trap
        trap += height[curr] - height[high];
      }
    }
  }
  // Return the total trapped rainwater
  return trap;
}

File: javascript/leet_code/42.mjs
/**
 * Calculates how much water can be trapped between the bars in a bar chart.
 * @param {number[]} height - An array of non-negative integers representing the heights of the bars.
 * @return {number} - The total amount of water that can be trapped.
 */
const trap = function (height) {
    // Initialize pointers and variables
    let left = 0, right = height.length - 1, lmax = 0, rmax = 0, water = 0;

    // Loop until the two pointers meet
    while (left < right) {
        // If the bar on the left is shorter or equal to the bar on the right
        if (height[left] <= height[right]) {
            // If the current bar is taller than the tallest bar we've seen from the left
            if (height[left] > lmax) {
                // Update the tallest bar from the left
                lmax = height[left];
            } else {
                // Add the difference between the tallest bar from the left and the current bar to the total
                water += lmax - height[left];
            }
            // Move the left pointer to the right
            left++;
        } else {
            // If the current bar is taller than the tallest bar we've seen from the right
            if (height[right] > rmax) {
                // Update the tallest bar from the right
                rmax = height[right];
            } else {
                // Add the difference between the tallest bar from the right and the current bar to the total
                water += rmax - height[right];
            }
            // Move the right pointer to the left
            right--;
        }
    }
    // Return the total amount of water that can be trapped
    return water;
};
